iceoryx_utils/test/moduletests/test_cxx_function_ref.cpp

```diff
@@ -0,0 +1,276 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "test.hpp"
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+constexpr int freeFuncTestValue = 42 + 42;
+constexpr int functorTestValue = 11;
+constexpr int memberFuncTestValue = 4273;
+
+int freeFunction()
+{
+    return freeFuncTestValue;
+}
+
+class Functor
+{
+  public:
+    int operator()()
+    {
+        return functorTestValue;
+    }
+};
+
+struct ComplexType
+{
+    char a;
+    int b;
+    float c;
+
+    bool operator==(const ComplexType& rhs) const
+    {
+        if (a == rhs.a && b == rhs.b && c == rhs.c)
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+};
+
+ComplexType returnComplexType(ComplexType foo)
+{
+    return foo;
+}
+
+class function_refTest : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+
+    void TearDown() override
+    {
+    }
+
+    int foobar()
+    {
+        return memberFuncTestValue;
+    }
+
+    uint8_t m_iter{0};
+};
+
+using function_refDeathTest = function_refTest;
+
+TEST_F(function_refTest, CreateEmptyIsFalse)
+{
+    function_ref<void()> sut;
+    EXPECT_FALSE(sut);
+}
+
+TEST_F(function_refDeathTest, CallEmptyLeadsToTermination)
+{
+    function_ref<void()> sut;
+    EXPECT_DEATH(sut(), ".*");
+}
+
+TEST_F(function_refTest, CreateValidByAssignIsTrue)
+{
+    auto lambda = [] {};
+    function_ref<void()> sut;
+    sut = lambda;
+    EXPECT_TRUE(sut);
+}
+
+TEST_F(function_refTest, CreateValidByAssignResultEqual)
+{
+    auto lambda = []() -> int { return 7253; };
+    function_ref<int()> sut;
+    sut = lambda;
+    EXPECT_THAT(sut(), Eq(7253));
+}
+
+TEST_F(function_refTest, CreateValidByCopyConstructResultEqual)
+{
+    auto lambda = []() -> int { return 3527; };
+    function_ref<int()> sut1{lambda};
+    function_ref<int()> sut2(sut1);
+    EXPECT_TRUE(sut2);
+    EXPECT_THAT(sut2(), Eq(3527));
+}
+
+TEST_F(function_refTest, CreateValidByCopyAssignResultEqual)
+{
+    auto lambda = []() -> int { return 43; };
+    function_ref<int()> sut2;
+    {
+        function_ref<int()> sut1{lambda};
+        EXPECT_THAT(sut1(), Eq(43));
+        EXPECT_FALSE(sut2);
+        sut2 = sut1;
+    }
+    EXPECT_THAT(sut2(), Eq(43));
+}
+
+TEST_F(function_refTest, CreateInvalidByCopyAssignIsFalse)
+{
+    auto lambda = []() -> int { return 44; };
+    function_ref<int()> sut2{lambda};
+    EXPECT_THAT(sut2(), Eq(44));
+    {
+        function_ref<int()> sut1;
+        EXPECT_FALSE(sut1);
+        sut2 = sut1;
+    }
+    EXPECT_FALSE(sut2);
+}
+
+TEST_F(function_refTest, CreateValidByMoveResultEqual)
+{
+    auto lambda = []() -> int { return 123; };
+    function_ref<int()> sut1{lambda};
+    function_ref<int()> sut2{std::move(sut1)};
+    EXPECT_TRUE(sut2);
```

Comment:
please make an ASSERT_THAT, else the access to sut2 further down might terminate the whole test

Reply:
Good point, will be done in #131